src/backend/cuda/Array.cpp

```diff
@@ -34,6 +34,18 @@ using std::shared_ptr;
 using std::vector;
 
 namespace cuda {
+
+template<typename T>
+void verifyTypeSupport() {
+    if ((std::is_same<T, double>::value || std::is_same<T, cdouble>::value) &&
+        !isDoubleSupported(getActiveDeviceId())) {
+        AF_ERROR("Double precision not supported", AF_ERR_NO_DBL);
+    } else if (std::is_same<T, common::half>::value &&
+               !isHalfSupported(getActiveDeviceId())) {
+        AF_ERROR("Half precision not supported", AF_ERR_NO_HALF);
+    }
+}
```

Comment:
I think we should move this to `src/backend/common` or `src/api/c` since this can be used from other locations too.

Also, Don't we need similar checks in `create*Array` helpers in others backends too ?

Reply:
This is already done in the OpenCL backend. It was not done in this backend because I intended to support half on older hardware and forgot to put it back in once I abandoned that idea.

Reply:
In that case, lets try to move this check into `src/api/c/` a more common location is perhaps `af_create_array` ?

Reply:
lots of functions create arrays. for example randu. I think implementing it in the common namespace is sufficient.

Reply:
Whichever location makes sense most.